File: WoofIOS/SwiftPackages/Domain/UseCase/Sources/UseCase/Favorites/ToggleDogFavoriteUseCase.swift
import Combine
import DI
import Models
import Repositories

@Factory
public final class ToggleDogFavoriteUseCase {
    private let favoritesRepository: FavoritesRepository

    public func callAsFunction(dog: Dog) -> AnyPublisher<Void, FavoritesError> {
        favoritesRepository.favoriteDogIds
            .first()
            .setFailureType(to: FavoritesError.self)
            .flatMap { [favoritesRepository] favoriteIds -> AnyPublisher<Void, FavoritesError> in
                switch favoriteIds.contains(dog.id) {
                case true: favoritesRepository.removeFavorite(dog: dog)
                case false: favoritesRepository.addFavorite(dog: dog)
                }
            }
            .eraseToAnyPublisher()
    }
}

File: WoofAndroid/domain/usecase/src/main/kotlin/com/perrystreet/woof/usecase/favorites/ToggleDogFavoriteUseCase.kt
package com.perrystreet.woof.usecase.favorites

import com.perrystreet.woof.models.dog.Dog
import com.perrystreet.woof.repositories.favorites.FavoritesRepository
import io.reactivex.rxjava3.core.Completable
import org.koin.core.annotation.Factory

@Factory
class ToggleDogFavoriteUseCase(
    private val favoritesRepository: FavoritesRepository,
) {
    operator fun invoke(dog: Dog): Completable =
        favoritesRepository.favoriteDogIds
            .firstOrError()
            .flatMapCompletable { favoriteIds ->
                when (dog.id in favoriteIds) {
                    true -> favoritesRepository.removeFavorite(dog)
                    false -> favoritesRepository.addFavorite(dog)
                }
            }
}
